File: neural_network/functions/mse_loss.py
class MSELoss:
    """Class to represent the mean squared error loss for regressional neural
    networks.
    """

    def __init__(self):
        """Constructor method
        """
        pass

    def __call__(self, y_hat: float, y: float) -> float:
        """The loss function.

        Parameters
        ----------
        y_hat : float
            Output value from neuron in output layer
        y : float
            Ground truth value

        Returns
        -------
        float
            Squared difference of the two values
        """
        return round((y_hat - y) ** 2, 8)

    def gradient(self, y_hat: float, y: float) -> float:
        """The gradient of the loss function.

        Parameters
        ----------
        y_hat : float
            Output value from neuron in output layer
        y : float
            Ground truth value

        Returns
        -------
        float
            Difference of the two values multiplied by 2
        """
        return round(2 * (y_hat - y), 8)

File: javaNeuralNetwork/src/main/java/neural_network/functions/MSELoss.java
package neural_network.functions;

/** Class to test the mean squared error loss for regression networks.
 *
 */
public class MSELoss {

    /** This method simply returns the squared error between the two values
     * {@code yHat} and {@code y}.
     *
     * @param yHat The output value from the {@code Neuron} in the output {@code Layer}.
     * @param y The true label for the datapoint.
     * @return The squared error between the two.
     */
    public static double call(double yHat, double y) {
        return (yHat - y) * (yHat - y);
    }

    /** This method returns the gradient of the loss function.
     *
     * @param yHat The output value from the {@code Neuron} in the output {@code Layer}.
     * @param y The true label for the datapoint.
     * @return The gradient of the squared error between the two.
     */
    public static double gradient(double yHat, double y) {
        return 2 * (yHat - y);
    }
}
